File: tutorial2-complete/src-kmp/app/src/main/kotlin/dev/modaal/foyer/app/AppHost.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.foyer.app

import dev.modaal.duet.shells.StoreHost
import dev.modaal.foyer.splash.SplashCompletionPath
import dev.modaal.foyer.splash.SplashDelegateEvent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

/** Which screen the app shows. Tutorial 3 replaces this with the root feature's state. */
sealed interface AppPhase {
  data object Splash : AppPhase

  data class Placeholder(val completedBy: SplashCompletionPath) : AppPhase
}

/**
 * The app's host at this step: it mounts the splash store on the retained
 * scope, receives the splash's delegate event, tears the splash down and
 * switches to the placeholder. Android-free, so the JVM test drives it.
 * Tutorial 3 replaces it with the root feature and its Builder.
 */
class AppHost(scope: CoroutineScope) {
  /** The teardown registry: what the host builds registers here and unwinds in reverse. */
  val host = StoreHost(scope)

  private val mutablePhase = MutableStateFlow<AppPhase>(AppPhase.Splash)
  val phase: StateFlow<AppPhase> = mutablePhase

  /** The splash store: mounted here, torn down when the splash completes. */
  var splash: SplashStore? =
    host.host(SplashBuilder().buildSplash(onDelegate = ::splashCompleted, scope = scope))
    private set

  private fun splashCompleted(event: SplashDelegateEvent) {
    // Both completion paths notify every time; the first notification moves
    // the app on and the rest arrive after the splash is gone.
    if (mutablePhase.value !is AppPhase.Splash) return
    val completed = event as SplashDelegateEvent.Completed
    splash?.teardown()
    splash = null
    mutablePhase.value = AppPhase.Placeholder(completedBy = completed.path)
  }

  /** Logical destruction only (finish, not rotation). */
  fun teardown() = host.teardownAll()
}

File: tutorial3-start/src-ios/App/Foyer/AppHost.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import FoyerKit
import Foundation
import SplashShell

/// Which screen the app shows. Tutorial 3 replaces this with the root feature's state.
enum AppPhase {
  case splash
  case placeholder(completedBy: SplashCompletionPath)
}

/// The app's host at this step: it mounts the splash, receives the splash's
/// delegate event, tears the splash down and switches to the placeholder.
/// Tutorial 3 replaces it with the root feature and its Builder.
@MainActor
final class AppHost: ObservableObject {
  @Published private(set) var phase: AppPhase = .splash

  /// The splash mount: built here, torn down when the splash completes.
  private(set) var splash: SplashChild?

  init() {
    splash = SplashBuilder().buildSplash { [weak self] event in
      self?.splashCompleted(event)
    }
  }

  /// The mount's bracket opens once the window is on screen.
  func activate() {
    splash?.shell.activate()
  }

  /// Logical destruction: the scene disconnected.
  func teardown() {
    splash?.shell.deactivate()
    splash = nil
  }

  private func splashCompleted(_ event: SplashDelegateEvent) {
    // Both completion paths notify every time; the first notification moves
    // the app on and the rest arrive after the splash is gone.
    guard case .splash = phase,
      case .completed(let completed) = onEnum(of: event)
    else { return }
    splash?.shell.deactivate()
    splash = nil
    phase = .placeholder(completedBy: completed.path)
  }
}
